### PR TITLE
V2 zones and od

### PR DESCRIPTION
**Pull request by e-kotov**
Hi, v2 zones and OD ready. Many functions unified with either `ver` argument or automatic version detection from dates.

```r
# Example usage
devtools::load_all()
Sys.setenv(SPANISH_OD_DATA_DIR = "your/data/folder")
zones <- "distr"

# v1 data (version detected from dates)
dates <- c("2020-02-14", "2020-03-14", "2021-02-14", "2021-02-14", "2021-02-15")
od_dist <- spod_get_od(zones = zones, dates = dates)

dplyr::glimpse(od_dist)

con <- od_dist$src$con
DBI::dbListTables(con)
DBI::dbDisconnect(con)


# v2 data (version detected from dates)
dates <- c("2022-02-14", "2022-03-14", "2023-02-14", "2023-02-14", "2024-02-15")
od_dist <- spod_get_od(zones = zones, dates = dates)

dplyr::glimpse(od_dist)
con <- od_dist$src$con
DBI::dbListTables(con)
DBI::dbDisconnect(con)


# zones
dist_v1 <- spod_get_zones("dist", ver = 1)
muni_v1 <- spod_get_zones("muni", ver = 1)
dist_v2 <- spod_get_zones("dist", ver = 2)
muni_v2 <- spod_get_zones("muni", ver = 2)
gau_v2 <- spod_get_zones("gau", ver = 2)

```

### CODE REVIEW

**Comment by e-kotov** [comment]
Using SQL text files massively simplified and improved robustness of the `spod_get_od()` function. So closes #22 .

**Comment by e-kotov** [comment]
I have not yet removed `spod_get()` because it is used in the main readme. Will have to take time to update the readme with the use of `spod_get_od()`.

**Comment by e-kotov** [comment]
Also, some sql files are in original Spanish (only the values for now, not the column or ENUM names). So there's potential of translating the tables to Spanish or any other language and to easily control that with some package level option via `options()`.

**Comment by Robinlovelace** [comment]
> I have not yet removed `spod_get()` because it is used in the main readme. Will have to take time to update the readme with the use of `spod_get_od()`.

Fine for the README to be out-of-date and happy to update it after this is merged. Or perhaps better, I could try updating the README as part of this PR so it's all done in one fell swoop.

**Comment by Robinlovelace** [comment]
Heads-up @e-kotov I'm hitting an error message after loading the package on this branch and running the following:

```r
dates <- seq(as.Date("2024-03-01"), as.Date("2024-03-07"), by = "day")
od_db <- spod_get_od(zones = "distritos", dates = dates)
```

**Comment by Robinlovelace** [comment]
```
> od_db <- spod_get_od(zones = "distritos", dates = dates)
Error: rapi_prepare: Failed to prepare query CREATE VIEW od_csv_raw AS SELECT *
/* csv_folder needs to be replaced with a valid path
in R use glue::glue() */
FROM read_csv_auto('/data/bronze/spanishoddata/raw_data_cache/v2//estudios_basicos/por-distritos/viajes/ficheros-diarios/**/*.csv.gz',
    delim='|',
    header=TRUE,
    hive_partitioning=TRUE,
    columns={
        'fecha': 'DATE',
        'periodo': 'INTEGER',
        'origen': 'VARCHAR',
        'destino': 'VARCHAR',
        'distancia': 'VARCHAR',
        'actividad_origen': 'VARCHAR',
        'actividad_destino': 'VARCHAR',
        'estudio_origen_posible': 'VARCHAR',
        'estudio_destino_posible': 'VARCHAR',
        'residencia': 'VARCHAR',
        'renta': 'VARCHAR',
        'edad': 'VARCHAR',
        'sexo': 'VARCHAR',
        'viajes': 'DOUBLE',
        'viajes_km': 'DOUBLE'
},
dateformat='%Y%m%d');
Error: IO Error: No files found that match the pattern "/data/bronze/spanishoddata/raw_data_cache/v2//estudios_basicos/
```

Any ideas? Seems like a path issue.

**Review by Robinlovelace — COMMENTED** [review]
Work in progress, see starter for updates to README in #26

**Comment by e-kotov** [comment]
> Heads-up @e-kotov I'm hitting an error message after loading the package on this branch and running the following:
> 
> ```r
> dates <- seq(as.Date("2024-03-01"), as.Date("2024-03-07"), by = "day")
> od_db <- spod_get_od(zones = "distritos", dates = dates)
> ```

does this work for you?

```r
od_db <- spod_get_od(zones = "distritos", dates = c(start = "2024-03-01", end = "2024-03-07"))
```

I think there's some bug with processing dates as `Dates`.

**Comment by Robinlovelace** [comment]
That fixes it, thanks Egor! As per commit above..

**Comment by e-kotov** [comment]
@Robinlovelace great, then I will look further into the date processing in a separate branch.

**Comment by Robinlovelace** [comment]
#28

**Comment by e-kotov** [comment]
> #29 FYI another issue I hit is in there.

should be fixed with this latest commit https://github.com/Robinlovelace/spanishoddata/pull/25/commits/557617d976ab7c9e61b5adb6928db7c5683e2dc0

I forgot to manually include the "externo" zone, which is the value for trips that don't have any zone in the spatial data (i.e. locations outside Spain, France, or Portugal).

**Comment by Robinlovelace** [comment]
Great fix, will merge #28 and keep the commits rolling on this branch.

File: NAMESPACE [hunk 1]
```diff
@@ -1,13 +1,8 @@
 # Generated by roxygen2: do not edit by hand
 
-export(spod_available_data_v1)
 export(spod_convert_od_v1_to_duckdb)
 export(spod_download_data)
 export(spod_get)
 export(spod_get_data_dir)
-export(spod_get_latest_v1_file_list)
```
Comment (Robinlovelace) [inline NAMESPACE:8]: Good by lots of exported functions with esoteric names 👍

File: NAMESPACE [hunk 2]
```diff
@@ -1,13 +1,8 @@
...
-export(spod_get_latest_v2_xml)
-export(spod_get_metadata)
-export(spod_get_od_v1)
+export(spod_get_od)
```
Comment (Robinlovelace) [inline NAMESPACE:7]: 👍

File: R/download_data.R [hunk 3]
```diff
@@ -2,12 +2,12 @@
...
 #' @param type The type of data to download. Can be `"origin-destination"` (or ust `"od"`), or `"trips_per_person"` (or just `"tpp"`) for v1 data. For v2 data `"overnight_stays"` (or just `"os"`) is also available. More data types to be supported in the future. See respective codebooks for more information. **ADD CODEBOOKS! to the package**
-#' @param zones The zones for which to download the data. Can be `"districts"` (or `"dist"`, `"distr"`, or the original Spanish `"distritos"`) or `"municipalities"` (or `"muni"`, `"municip"`, or the original Spanish `"municipios"`). Additionaly, these can be `"large_urban_areas"` (or `"lau"`, or the original Spanish `"grandes_areas_urbanas"`, or `"gau"`) for v2 data.
+#' @param zones The zones for which to download the data. Can be `"districts"` (or `"dist"`, `"distr"`, or the original Spanish `"distritos"`) or `"municipalities"` (or `"muni"`, `"municip"`, or the original Spanish `"municipios"`) for both data versions. Additionaly, these can be `"large_urban_areas"` (or `"lau"`, or the original Spanish `"grandes_areas_urbanas"`, or `"gau"`) for v2 data (2022 onwards).
```
Comment (Robinlovelace) [inline R/download_data.R:5]: Great concise description of data availability for v2 (most data).

File: R/duckdb_helpers.R [hunk 4]
```diff
@@ -48,84 +49,76 @@
...
+    stop("The 'gau' zones are not available in v1 data.")
+  }
 
-  # create view of csv files and preset variable types
-  if (zones == "distritos") {
-    DBI::dbSendStatement(
-      con,
-      dplyr::sql(
-        glue::glue(
-          "CREATE VIEW od_csv_raw AS SELECT *
```
Comment (Robinlovelace) [inline R/duckdb_helpers.R:78]: Sooo much better with this as .sql files, right?

File: R/duckdb_helpers.R [hunk 5]
```diff
@@ -136,67 +129,91 @@ spod_duckdb_od_v1 <- function(
...
-    DBI::dbSendStatement(
-      con,
-      dplyr::sql("CREATE TYPE ACTIV_ENUM AS ENUM ('home', 'work', 'other')")
-    )
+  # create ACTIV_ENUM (for all except for municipalities in v1 data)
+  if ( !( zones == "municipios" & ver == 1) ) {
+    # LOAD SQL STATEMENT
+    sql_create_activ_enum <- readLines(
+      system.file(
+        glue::glue("extdata/sql-queries/v{ver}-od-enum-activity-en.sql"),
```
Comment (Robinlovelace) [inline R/duckdb_helpers.R:137]: Nice!

File: R/get_v1_data.R [hunk 6]
```diff
@@ -273,28 +290,30 @@ spod_clean_zones_v1 <- function(zones_path) {
...
     zones = c(
       "districts", "dist", "distr", "distritos",
       "municipalities", "muni", "municip", "municipios"
     ),
     dates = NULL,
     data_dir = spod_get_data_dir(),
     quiet = FALSE,
-    duck_max_mem = 2,
-    duck_max_threads = parallelly::availableCores())
+    duck_max_mem = 3,
```
Comment (Robinlovelace) [inline R/get_v1_data.R:309]: Why 3 here out of interest?

File: R/get_v1_data.R [hunk 7]
```diff
@@ -273,28 +290,30 @@ spod_clean_zones_v1 <- function(zones_path) {
...
+    duck_max_threads = parallelly::availableCores() - 1)
```
Comment (Robinlovelace) [inline R/get_v1_data.R:310]: 👍 to not max out machines!

**Review by Robinlovelace — APPROVED** [review]
Good to go from my perspective, feel free to make any further tweaks and merge /tidy at your leisure @e-kotov

File: R/duckdb_helpers.R (hunk 4 above)
Comment (e-kotov) [reply R/duckdb_helpers.R:78]: Absolutely! And as I said, I love it how it made the internal helper function to create duckdb table "views" to csv folders much more concise. That was a great suggestion, @Robinlovelace !

File: R/get_v1_data.R (hunk 6 above)
Comment (e-kotov) [reply R/get_v1_data.R:309]: @Robinlovelace tested with 2GB and it failed. I did not test it much beyond 7-10 files, we will see later how it behaves with more files or during conversion. In the HPC I'm used to giving it at least 30GB to convert the complete dataset. DuckDB is clever enough to [disk spilling](https://duckdb.org/2024/07/09/memory-management.html#:~:text=DuckDB%20deals%20with%20these%20scenarios%20by%20disk%20spilling) but in our case this 3GB seems like the bare minimum that it wants to be allocatable in RAM.

File: R/get_v1_data.R (hunk 7 above)
Comment (e-kotov) [reply R/get_v1_data.R:310]: yep, to keep one core to prevent the OS from freezing

**Comment by e-kotov** [comment]
Great, I will merge now. Will deal with the dates a bit later this week #27 .
Closes #22 .

File: R/get_v1_data.R (hunk 6 above)
Comment (Robinlovelace) [reply R/get_v1_data.R:309]: OK makes sense. Agree at some point may be worth dynamically setting this.

**Comment by e-kotov** [comment]
Also closes:
https://github.com/e-kotov/mitma-data-issues/issues/2
https://github.com/e-kotov/mitma-data-issues/issues/3
https://github.com/e-kotov/mitma-data-issues/issues/4

**Comment by e-kotov** [comment]
Also closes:
https://github.com/e-kotov/mitma-data-issues/issues/5
https://github.com/e-kotov/mitma-data-issues/issues/6
https://github.com/e-kotov/mitma-data-issues/issues/7

**Pull request merged**